### ISSUE TITLE
The first step in "Flow" is execute twice

### ISSUE THREAD

**Issue by bio-engineer** [spring-projects/spring-batch#4432]
**Bug description**
If you create a "Flow" and use the "next" method instead of the "start" method to add the first step, then this step will be executed twice.

The problem is in the line `addTransition("COMPLETED", next);`. It is she who makes Flow perform the step a second time.

https://github.com/spring-projects/spring-batch/blob/e6c27273fa2b3713c6f2d472bf3de1b18f8e5eba/spring-batch-core/src/main/java/org/springframework/batch/core/job/builder/FlowBuilder.java#L249

Current version:
```java

	private void doNext(Object input) {
		if (this.currentState == null) {
			doStart(input);
		}
		State next = createState(input);
		addTransition("COMPLETED", next);
		addTransition("*", failedState);
		this.currentState = next;
	}

```

Fixed version:
```java

	private void doNext(Object input) {
		if (this.currentState == null) {
			doStart(input);
		} else {
			State next = createState(input);
			addTransition("COMPLETED", next);
                        addTransition("*", failedState);
			this.currentState = next;
		}
		
	}

```

**Environment**
```
	<groupId>org.springframework.boot</groupId>
 	<artifactId>spring-boot-starter-batch</artifactId>
 	<version>3.1.2</version>
```

**Steps to reproduce**
```java
	FlowBuilder<Flow> flowBuilder = new FlowBuilder<>("someFlowName");
	flowBuilder.next(step);
	return flowBuilder.build();
```

**Expected behavior**
Created "Flow" with one step. Step executed once.

**Minimal Complete Reproducible example**
The step was to transfer data from one table to another identical table. The second run of the step failed only because a unique primary key constraint was created.
![image](https://github.com/spring-projects/spring-batch/assets/65017238/1ee106cb-2022-4a4e-aa85-4ccf63e491be)

**Comment by fmbenhassine** [comment]
That's a valid issue. Thank you for raising it and for providing a patch!

When I tried to reproduce the issue, I noticed that the job never finishes (see attached minimal project, based on the [issue reporting project template](https://github.com/spring-projects/spring-batch/blob/main/ISSUE_REPORTING.md)).

You are welcome to contribute the patch in a PR if you want. Thank you.

---

Minimal complete example:

[gh4432.zip](https://github.com/spring-projects/spring-batch/files/12897238/gh4432.zip)

**Comment by injae-kim** [comment]
Can I handle this issue? I'll create PR soon~! 🙇

**Comment by injae-kim** [comment]
> https://github.com/spring-projects/spring-batch/issues/4432#issuecomment-1761975340
> When I tried to reproduce the issue, I noticed that the **job never finishes**

```
builder.next(createCompleteStep("stepA"))
       .end()
       .start();
--
[main] INFO org.springframework.batch.core.job.SimpleStepHandler - Executing step: [stepA]
[main] INFO org.springframework.batch.core.job.SimpleStepHandler - Executing step: [stepA]
[main] INFO org.springframework.batch.core.job.SimpleStepHandler - Executing step: [stepA]
... "stepA" never finishes
```

Hi @fmbenhassine, I also tried to reproduce this issue and check job never finishes too!
So I make fix PR #4475. PTAL 🙇 thanks!

**Comment by bio-engineer** [comment]
@fmbenhassine Good afternoon
I created a "Pull Request" with a minimum number of changes that correct the problem described above.

Added a test that shows that the code I added does indeed fix the error
